File: sdk/chocobase-py/src/chocobase/storage.py
from typing import Dict, Any, Optional

class StorageClient:
    """Client for Object Storage buckets and files."""

    def __init__(self, url: str, key: str):
        self.url = f"{url}/v1/storage/v1"
        self.key = key

    def from_(self, bucket_id: str) -> "BucketClient":
        return BucketClient(self.url, bucket_id, self.key)

class BucketClient:
    def __init__(self, base_url: str, bucket_id: str, key: str):
        self.url = f"{base_url}/object/{bucket_id}"
        self.bucket_id = bucket_id
        self.key = key

    def upload(self, path: str, file_bytes: bytes, content_type: str = "application/octet-stream") -> Dict[str, Any]:
        return {
            "data": {"path": f"{self.bucket_id}/{path}"},
            "error": None,
        }

    def create_signed_url(self, path: str, expires_in: int = 3600) -> Dict[str, Any]:
        return {
            "data": {
                "signed_url": f"{self.url}/sign/{path}?expires_in={expires_in}"
            },
            "error": None,
        }

File: sdk/chocobase-dart/lib/src/storage.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;

class StorageResponse {
  final dynamic data;
  final String? error;

  StorageResponse({this.data, this.error});
}

class StorageFileApi {
  final String baseUrl;
  final String bucket;
  final Map<String, String> headers;

  StorageFileApi(this.baseUrl, this.bucket, this.headers);

  Future<StorageResponse> upload(String path, Uint8List fileBytes, {String contentType = 'application/octet-stream'}) async {
    final uri = Uri.parse('$baseUrl/v1/storage/v1/object/$bucket/$path');
    final uploadHeaders = Map<String, String>.from(headers);
    uploadHeaders['Content-Type'] = contentType;

    final response = await http.post(uri, headers: uploadHeaders, body: fileBytes);
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return StorageResponse(data: jsonDecode(response.body));
    } else {
      return StorageResponse(error: 'Upload failed: ${response.statusCode}');
    }
  }

  Future<String?> createSignedUrl(String path, {int expiresIn = 3600}) async {
    final uri = Uri.parse('$baseUrl/v1/storage/v1/object/sign/$bucket/$path');
    final response = await http.post(
      uri,
      headers: headers,
      body: jsonEncode({'expires_in': expiresIn}),
    );

    if (response.statusCode >= 200 && response.statusCode < 300) {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final signedUrl = json['signed_url'] as String?;
      return signedUrl != null ? '$baseUrl$signedUrl' : null;
    }
    return null;
  }
}

class StorageClient {
  final String baseUrl;
  final Map<String, String> headers;

  StorageClient(this.baseUrl, this.headers);

  StorageFileApi from(String bucket) {
    return StorageFileApi(baseUrl, bucket, headers);
  }
}
